directorywatcher/directorywatcher.go

```diff
@@ -0,0 +1,288 @@
+package directorywatcher
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sync/atomic"
+
+	"gopkg.in/fsnotify.v1"
+
+	"github.com/reddit/baseplate.go/filewatcher"
+	"github.com/reddit/baseplate.go/internal/limitopen"
+	"github.com/reddit/baseplate.go/log"
+)
+
+// DirectoryWatcher loads and parses data from a file and watches for changes to that
+// file in order to refresh it's stored data.
+type DirectoryWatcher interface {
+	// Get returns the latest, parsed data from the DirectoryWatcher.
+	Get() interface{}
+
+	// Stop stops the DirectoryWatcher.
+	//
+	// After Stop is called you won't get any updates on the file content,
+	// but you can still call Get to get the last content before stopping.
+	//
+	// It's OK to call Stop multiple times.
+	// Calls after the first one are essentially no-op.
+	Stop()
+}
+
+// AddFile is a type of function that should be ran to handle
+// adding data after a file has been parsed
+type AddFile func(d interface{}, file interface{}) (data interface{}, err error)
+
+// RemoveFile is a type of function that should be ran to handle removing
+// data after a file has been removed from the watcher
+type RemoveFile func(d interface{}, path string) (data interface{}, err error)
```

Comment:
I assume an user case would be that we split a huge json file into smaller ones, and the `AddFile` callback is supposed to stitch the json data together into a single object. assuming we have:

```json
// foo.json
{
  "foo": {
    "a": 1,
    "b": 2,
  }
}

// bar.json
{
  "bar": {
    "a": 3,
    "b": 4,
  }
}
```

so we want the stitched together object be something like:
```go
map[string]object{
  "foo": object{
    A: 1,
    B: 2,
  },
  "bar": object{
    A: 3,
    B: 4,
  },
}
```

Now `bar.json` is removed, `RemoveFile` is called with `d` being the whole object and path being `bar.json`. But how does the implementation know that it should remove the `bar` key from the object? The args of `RemoveFile` and `AddFile` are not symmetry, so it has no knowledge on the association between the path and the actual data to remove.

I'm also questioning the decision to make it recursive. From the currently defined interface we are treating all the files equally (e.g. there's no difference between "<root>/foo.json" and "<root>/foo/bar.json" if they have the same content), but if we treat all files equally what's the point of having directory hierarchy?

Reply:
So from the perspective of the Vault CSI plugin, a files position in a directory would imply a difference in data since it mirrors how secrets are stored in vault. Each secret will be its own file.

I do agree with the assymetry of the add and remove functions being weird so I may take Kyle's advice and ditch the remove function entirely